Add error handling and some cleanup to ns-build.sh

groboclown opened the pull request:
The bash script includes many parts that, if an error occurs, would cause the script to not fail correctly, or, even worse, leave the application in an unexpected state.

This improvement starts by introducing a cleanup function that runs on exit (`trap cleanup EXIT`), so that if the user interrupts the execution, the script correctly cleans up after itself.

Because the cleanup script has an expectation for the directory it runs in, it means the 'cd' commands could throw off its expected execution directory.  To this end, the 'cd' only runs in a sub-shell for the zip commands.

Next, most commands include an exit-on-failure ('|| exit 1').  This allows the script to better handle environment or build issues.

This also includes some extra cleanup.  The app-backup and nodejs and standalone directories are now put inside the new .ns-build directory.  This should help make .gitignore easier, and can help reduce root directory clutter and possible interference with the application structure.  Some bash-specific code changed to more sh-friendly syntax (such as `[[` to `[`), and included extra argument quoting.

The one line that seems odd is the `KEY="$(echo "$ARGUMENT" | cut -f1 -d=)"` - the '-f1' would take all fields before the '-d=' mark, but I'm not sure if that's the intended behavior; should this instead be a '-f2-'?

With a bit of extra effort, the argument parsing loop could be turned into a more general SH script instead of bash-specific, but I'll leave to that to another day.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- **Chores**
	- Updated the `.gitignore` file to enhance file and directory exclusions, including new patterns for `.ns-build`, `.zip`, deployments, and standalone directories.
	- Improved the `ns-build.sh` script for better error handling and functionality, including a cleanup process and directory structure adjustments.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

groboclown committed:
Add error handling to ns-build.sh

The bash script includes many parts that, if an error occurs, would cause the script to not fail correctly, or, even worse, leave the application in an unstable state.

This starts by introducing a cleanup function that runs on exit ('trap cleanup EXIT'), so that if the user interrupts the execution, the script correctly cleans up after itself.

Because the cleanup script has an expectation for the directory it runs in, it means the 'cd' commands could throw off its expected execution directory.  To this end, the 'cd' only runs in a sub-shell for the zip commands.

Next, most commands include an exit-on-failure ('|| exit 1').  This allows the script to better handle environment or build issues.

This also includes some extra cleanup.  The app-backup and nodejs and standalone directories are now put inside the new .ns-build directory.  This should help make .gitignore easier, and can help reduce root directory clutter and possible interference with the application structure.  Some bash-specific code changed to more sh-friendly syntax (such as '[[' to '['), and included extra argument quoting.

The one line that seems odd is the 'KEY="$(echo "$ARGUMENT" | cut -f1 -d=)" - the '-f1' would take all fields before the '-d=' mark, but I'm not sure if that's the intended behavior; should this instead be a '-f2-'?

With a bit of extra effort, the argument parsing loop could be turned into a more general SH script instead of bash-specific, but I'll leave to that to another day.

emyriounis commented on packages/ns-build/bin/ns-build.sh:
```diff
@@ -18,83 +18,100 @@ do
+
+# Keep necessary files
+cp -a .next/static .next/standalone/.next || exit 1
+cp -a .next/standalone .ns-build/standalone || exit 1
```
one of the child modules expects the `standalone` folder in the root directory

please update [this too](https://github.com/emyriounis/terraform-aws-nextjs-serverless/blob/177458ff35359baca203441b38a30c3f4752a4e5/modules/static-assets-hosting/main.tf#L25), or move the folder back to the root directory

groboclown replied:
Good catch!  I'll get that patched up.

emyriounis reviewed (changes requested):
great work, please check my note

groboclown committed:
Keep standalone as a top-level directory

The 'standalone' directory should live in the base directory, not as a build artifact, due to how the lambda references the standalone directory.

This does mean that users of the tool should not create the standalone directory or file, because the build will delete it.

groboclown committed:
Re-add the standalone directory to the gitignore file.

emyriounis commented:
@groboclown I tested your changes & i get error 500, please test yourself too to identify & fix the issue

the lambda error:
<img width="1099" alt="Screenshot 2024-07-24 at 14 48 08" src="https://github.com/user-attachments/assets/b3dfce7a-1f6a-408b-85bb-be4f12975bc7">

emyriounis commented:
@coderabbitai review
